compressors/arithmetic_coding.py

```diff
@@ -90,6 +90,70 @@ def update_model(self, s):
                 self.freqs_current.freq_dict[s] = max(f // 2, 1)
 
 
+class AdaptiveOrderKFreqModel(FreqModelBase):
+    """kth order adaptive frequency model.
+
+    Parameters:
+        alphabet: the alphabet (provided as a list)
+        k:        the order, k >= 0 (kth order means we use past k to predict next, k=0 means iid)
+
+    """
+
+    def __init__(self, alphabet: List, k: int, max_allowed_total_freq):
+        assert k >= 0
+        self.k = k
+        # map alphabet to index from 0 to len(alphabet) so we can use with numpy array
+        self.alphabet_to_idx = {alphabet[i]: i for i in range(len(alphabet))}
+        # keep freq/counts of (k+1) tuples, initialize with all 1s (uniform)
+        self.freqs_kplus1_tuple = np.ones([len(alphabet)] * (k + 1), dtype=int)
+        self.max_allowed_total_freq = max_allowed_total_freq
+        # keep track of past k symbols (i.e., alphabet index) seen. Initialize with all 0s.
+        self.past_k = [0] * k
```

Comment:
So here we are assuming that when we start calculating adaptive frequencies, we start with past context of all k symbols seen to be the 0th alphabet index symbol, correct? how do typical encoders handle this case for the initialization?

Reply:
yes, that's it. I added a comment. I think generally you need to do something or the other for the first few. I chose a simple method to avoid complications.